### ISSUE TITLE
Make custom fields repeatable

### ISSUE THREAD

**Issue by GoogleCodeExporter** [viktorthang/wordpress-custom-content-type-manager#79]
```
What functionality do you want to see?

I would love to make custom fields repeatable.  For example, upload any number 
of images to a Product content type.  Or, have many phone numbers & contact 
details text boxes for a Company content type.

Do you have examples of similar functionality on other sites or in other
plugins?

Yes, the simple-fields plugin, though I haven't actually used it, the 
screenshots show that they have the right idea:
http://wordpress.org/extend/plugins/simple-fields/screenshots/
When defining your fields, you can tick the "repeatable" checkbox.  Then when 
editing that content type, there is a '+' button to add another copy of the 
field.

Please provide any additional information below.

I'm not too familiar with wordpress internals - is this possible without 
changing the database etc?  I'd be interested to see how the "simple-fields" 
plugin handles it.

I think you could probably make the interface a little classier than they've 
done as well.  I'd be happy to lend a hand with mocking up a nice UI to make 
things repeatable.

```

Original issue reported on code.google.com by `jason.on...@gmail.com` on 5 May 2011 at 5:21

**Comment by GoogleCodeExporter** [comment]
```
The problem here is with the  architecture: supporting a finite number is 
always easier than supporting a number of n.  Some of how WP handles this 
natively is fatally flawed from an architectural point of view, but I think I 
could do something like this using the multi-select.  I have to think about the 
"correct" way to implement this.

I am planning on having a front-end UI / UX guy groom through this thing before 
version 1.0.  There are a lot of pieces  that have to work together, and I've 
given precedence to functionality first.  CSS and all that will get cleaned up 
later.
```

Original comment by `fireproofsocks` on 5 May 2011 at 7:17
- Changed state: **InReview**
- Added labels: **Type-Enhancement**

**Comment by GoogleCodeExporter** [comment]
```
Hmm... looking at the "Simple Fields" plugin, I'm not convinced that's the 
right approach architecturally.  Hard to explain, but my focus in this plugin 
is to give the post-type (aka content-type) the priority, but the Simple Fields 
plugin is giving precedence to the custom field as the unit.  It'd be nice to 
have a more normalized data set for this, but the UI would get more complex 
with having to define the relations required to support the normalized 
structure.  I'm pretty happy with how the simplified approach I've taken allows 
easy importing and exporting of the data definitions -- the drawback of the 
current architecture is that you have to recreate custom field defs, which 
could be a pain if you added the same type of field to multiple post-types.

So there is an option here to completely normalize the custom field definitions 
-- this would allow for easier porting of custom fields between post-types...  
doing that would require re-writing the data structure yet again, but hopefully 
it could be done without as much trouble as the 0.8.9 update caused.  If I'm 
gonna do that, it'd have to be done prior to version 1.0... after that, I want 
the internals of this to be settled.

All that normalization talk aside, the debate is still open as to how to 
support multiple values for a single field instance.  I could use a checkbox 
for something like "support multiple" that would allow users to create multiple 
instances of the field in their posts.  That would require opening up 
array-type output filters for any custom field that supported a "multi" option.

Still thinking about this one...
```

Original comment by `fireproofsocks` on 15 May 2011 at 2:22

**Comment by GoogleCodeExporter** [comment]
```
Assuming we have a normalized storage structure, how would this even work?  
Each custom field type class has implemented the get_edit_field_instance() 
function, which draws a form for a single instance of the custom field; adding 
an array of instances would require custom javascript, changing the "name" 
attributes to support arrays, e.g. name="my_field[]".

Maybe I can add a method to the class for "get_edit_instance_repeatable()" or 
something so the name etc. can be adjusted and valid Javascript can be added.

Output filters... e.g .for images, you'd want an array of src's...  is it 
possible to have an array version of any Output filter?
```

Original comment by `fireproofsocks` on 18 Aug 2011 at 6:47

**Comment by GoogleCodeExporter** [comment]
Original comment by `fireproofsocks` on 29 Sep 2011 at 5:54
- Changed state: **Accepted**
- Added labels: **Milestone-Release0.9.5**, **Priority-Medium**

**Comment by GoogleCodeExporter** [comment]
```
Committed revision 461087.

This will be in 0.9.5.  The use of the Ajax API and a better implementation of 
the thickbox have made all kinds of good things possible here.  The most 
important repeatable fields will be the relation fields (relations, images, 
media).

The relation fields have been implemented on the current dev version and the 
other fields are to follow and will be released in 0.9.5.
```

Original comment by `fireproofsocks` on 9 Nov 2011 at 5:37
- Changed state: **Fixed**

**Comment by GoogleCodeExporter** [comment]
```
Issue 239 has been merged into this issue.
```

Original comment by `fireproofsocks` on 23 Nov 2011 at 6:41

**Comment by GoogleCodeExporter** [comment]
```
Any chance that we can set a limit on repeatable fields. In some other CMS I 
work with, there is a field where you can add an image and then another "select 
file" type field shows up below, allowing you to add more. In the back end you 
can set it to a set #, limiting a user to only adding X images and no more than 
that. Or, they can choose "unlimited" which will allow someone to add and add 
forever and ever.  Any chance we can modify the "repeatable" checkbox to also 
allow us to set a specific number limit?
```

Original comment by `1982mary...@gmail.com` on 10 Apr 2013 at 4:59

**Comment by GoogleCodeExporter** [comment]
```
Good idea: this should be added as a separate feature request.
```

Original comment by `ever...@fireproofsocks.com` on 14 Apr 2013 at 6:29
